javalib/src/main/scala/java/util/concurrent/ConcurrentLinkedQueue.scala

```diff
@@ -0,0 +1,45 @@
+package java.util.concurrent
+
+import java.util._
+
+class ConcurrentLinkedQueue[E]()
+    extends AbstractQueue[E] with Queue[E] with Serializable {
+
+  def this(c: Collection[_ <: E]) = {
+    this()
+    addAll(c)
+  }
+
+  private val inner: LinkedList[E] = new LinkedList()
+
+  override def add(e: E): Boolean =
+    if (e == null) throw new NullPointerException()
+    else inner.add(e)
+
+  override def offer(e: E): Boolean =
+    add(e)
+
+  override def poll(): E =
+    inner.poll()
+
+  override def peek(): E =
+    inner.peek()
+
+  override def isEmpty(): Boolean =
+    inner.isEmpty()
+
+  override def size(): Int =
+    inner.size()
+
+  override def contains(o: Any): Boolean =
+    inner.contains(o)
+
+  override def remove(o: Any): Boolean =
+    inner.remove(o)
+
+  override def addAll(c: Collection[_ <: E]): Boolean =
+    inner.addAll(c)
+
+  override def iterator(): Iterator[E] =
+    inner.iterator()
```

Comment:
I just figure out that the Iterator on ConcurrentLinkedQueue does not guarantee as said:
```guarantees to traverse elements as they existed upon construction of the iterator```
in fact:
```
scala> val queue= new ConcurrentLinkedQueue[Int]()
queue: java.util.concurrent.ConcurrentLinkedQueue[Int] = []

scala> queue.add(1)
res6: Boolean = true

scala> queue.add(2)
res7: Boolean = true

scala> queue.add(3)
res8: Boolean = true

scala> val iter = queue.iterator
iter: java.util.Iterator[Int] = java.util.concurrent.ConcurrentLinkedQueue$Itr@c10d272

scala> queue.clear

scala> iter.next
res10: Int = 1

scala> iter.remove

scala> iter.next
java.util.NoSuchElementException
  at java.util.concurrent.ConcurrentLinkedQueue$Itr.next(ConcurrentLinkedQueue.java:732)
  ... 43 elided
```

looks like that a "next element caching" can exactly reproduce the behaviour (and also satisfy your tests...)